Add AI Examples (#995)
edgio-docs/edgio-docs: src/guides/v7/ai/langchain.md

---
title: LangChain
---

> This template scaffolds a LangChain.js + Next.js starter app and showcases several use cases for different LangChain modules.

Reviewer: Revise this sentence to explain how Edgio fits into this workflow.

    const messages = body.messages ?? [];
    const formattedPreviousMessages = messages.slice(0, -1).map(formatMessage);
    const currentMessageContent = messages[messages.length - 1].content;
    const prompt = PromptTemplate.fromTemplate(TEMPLATE);

    const model = new ChatOpenAI({
      temperature: 0.8,
      modelName: 'gpt-3.5-turbo-1106',
    });

    const outputParser = new HttpResponseOutputParser();

    const chain = prompt.pipe(model).pipe(outputParser);

    const stream = await chain.stream({
      chat_history: formattedPreviousMessages.join('\n'),
      input: currentMessageContent,
    });

    return new StreamingTextResponse(stream);
  } catch (e: any) {
    return NextResponse.json({error: e.message}, {status: 500});
  }
}
```

- The `POST` handler initializes and calls a simple chain with a prompt, chat model, and output parser. [See the LangChain docs](https://js.langchain.com/docs/guides/expression_language/cookbook#prompttemplate--llm--outputparser) for more information
- Chat models stream message chunks rather than bytes, so output parsing is handling serialization and byte-encoding by initializing `HttpResponseOutputParser()` to `outputParser`.

> - To select a different model, the client would be initialized like so (see a [full list of supported models here](https://js.langchain.com/docs/modules/model_io/models/)):

Reviewer: ```suggestion
Select a different model by initializing a client as shown below. [View a list of supported models.](https://js.langchain.com/docs/modules/model_io/models/)
```

```ts
import {ChatAnthropic} from 'langchain/chat_models/anthropic';
const model = new ChatAnthropic({});
```

> - Instead of using `prompt.pipe()`, you can initialize `chain` like so:

Reviewer: Is it obvious why they would want to initialize `chain` instead of using `prompt.pipe()`. If not, explain why they would want to do that. Also, this line should not be a bullet point.

Author: This is a LangChain specific thing, they have a DSL, [LangChain Expression Language](https://js.langchain.com/docs/expression_language) and an abstraction called [Chains](https://js.langchain.com/docs/modules/chains/) for sequences multiple calls to LLMs and other APIs. It's probably not worth getting into cause they have docs on these topics so I'll remove this part.

```ts
import {RunnableSequence} from 'langchain/schema/runnable';
const chain = RunnableSequence.from([prompt, model, outputParser]);
```

### Agents and Retrieval {/* agents-and-retrieval */}

> To try out the Agent or Retrieval examples, follow the instructions in the [Edgio AI repository](https://github.com/Edgio/edgio-ai).

Reviewer: Why do we link out to these examples instead of documenting them here?

Author: The template is setup so you can run it with just an OpenAI key and use the Chat tab with the rest of the tabs as additional steps that require other API keys and eventually a vector database.

- The thought was just have the Quick Start experience in the docs and if you want to configure it further for a more advanced use case you could go to the instructions in the repo there.
- If you try to setup all the keys before you start interacting with the chat it feels like a ton of work until you get to any payoff.
- This also keeps it comparable to the other two examples in case anyone wanted to try running all three.
- But I agree it could also make sense to have the full instructions in the docs and if we wanted them there I could just copy the README that's in the repo into the docs:

https://github.com/edgio-docs/edgio-v7-ai-example/tree/main/langchain/README.md
